File: src/numericalAnalysis/MeshSurfaceVisitor.cpp
#include "java/util/ArrayList.txx"

#include "numericalAnalysis/MeshSurfaceVisitor.h"
#include "numericalAnalysis/PatchVisitor.h"

/**
Initializes MeshSurface/Patch links with safe defaults without running numerical analysis.
*/
void
MeshSurfaceVisitor::initializeFacesDefaults(MeshSurface *mesh) {
    if ( mesh == nullptr ) {
        return;
    }
    for ( int i = 0; mesh->faces != nullptr && i < mesh->faces->size(); i++ ) {
        Patch *face = mesh->faces->get(i);
        if ( face == nullptr ) {
            continue;
        }
        face->material = mesh->material;
    }
}

/**
Fills in the MeshSurface back pointer of the face belonging to the given surface
*/
void
MeshSurfaceVisitor::surfaceConnectFace(MeshSurface *mesh, Patch *face) {
    int i;

    face->material = mesh->material;

    // Also fill in a nicer default color for the patch
    switch ( mesh->colorFlags ) {
        case MaterialColorFlags::FACE_COLORS:
            break;
        case MaterialColorFlags::VERTEX_COLORS:
            // Average color of the vertices
            face->color.set(0, 0, 0);
            for ( i = 0; i < face->numberOfVertices; i++ ) {
                face->color.r += face->vertex[i]->color.r;
                face->color.g += face->vertex[i]->color.g;
                face->color.b += face->vertex[i]->color.b;
            }
            face->color.r /= static_cast<float>(i);
            face->color.g /= static_cast<float>(i);
            face->color.b /= static_cast<float>(i);
            break;
        default: {
            ColorRgb rho;
            rho = PatchVisitor::averageNormalAlbedo(face, BRDF_DIFFUSE_COMPONENT | BRDF_GLOSSY_COMPONENT);
            rho.set(face->color.r, face->color.g, face->color.b);
        }
    }
}

/**
Fill in the MeshSurface back pointer of the FACEs in the MeshSurface
*/
void
MeshSurfaceVisitor::fillFacesBackPointers(MeshSurface *mesh) {
    initializeFacesDefaults(mesh);
    for ( int i = 0; mesh->faces != nullptr && i < mesh->faces->size(); i++ ) {
        MeshSurfaceVisitor::surfaceConnectFace(mesh, mesh->faces->get(i));
    }
}

File: typescript/src/main/vsdk/toolkit/numericalAnalysis/MeshSurfaceVisitor.ts
import { ColorRgb } from "../common/color/ColorRgb";
import { BsdfComponent } from "../material/BsdfComponent";
import { MaterialColorFlags } from "../skin/MaterialColorFlags";
import { MeshSurface } from "../skin/MeshSurface";
import { Patch } from "../skin/Patch";
import { PatchVisitor } from "./PatchVisitor";

export class MeshSurfaceVisitor {
  private static surfaceConnectFace(mesh: MeshSurface, face: Patch): void {
    face.material = mesh.material;

    switch (MeshSurface.colorFlags) {
      case MaterialColorFlags.FACE_COLORS:
        break;
      case MaterialColorFlags.VERTEX_COLORS: {
        face.color.set(0, 0, 0);
        let i = 0;
        for (; i < face.numberOfVertices; i++) {
          const vertex = face.vertex[i];
          if (vertex === null) {
            continue;
          }
          face.color.r += vertex.color.r;
          face.color.g += vertex.color.g;
          face.color.b += vertex.color.b;
        }
        if (i > 0) {
          face.color.r /= i;
          face.color.g /= i;
          face.color.b /= i;
        }
        break;
      }
      default: {
        const rho: ColorRgb = PatchVisitor.averageNormalAlbedo(
          face,
          BsdfComponent.BRDF_DIFFUSE_COMPONENT | BsdfComponent.BRDF_GLOSSY_COMPONENT
        );
        rho.set(face.color.r, face.color.g, face.color.b);
      }
    }
  }

  public static initializeFacesDefaults(mesh: MeshSurface | null): void {
    if (mesh === null) {
      return;
    }
    for (let i = 0; mesh.faces !== null && i < mesh.faces.length; i++) {
      const face = mesh.faces[i];
      if (face === null) {
        continue;
      }
      face.material = mesh.material;
    }
  }

  public static fillFacesBackPointers(mesh: MeshSurface): void {
    MeshSurfaceVisitor.initializeFacesDefaults(mesh);
    for (let i = 0; mesh.faces !== null && i < mesh.faces.length; i++) {
      const face = mesh.faces[i];
      if (face !== null) {
        MeshSurfaceVisitor.surfaceConnectFace(mesh, face);
      }
    }
  }
}
